### PR TITLE
add rescue on kill tile

### PR DESCRIPTION
**Pull request by ewancg**
Some people were tired of having to re-enter practice mode, teleport back to where they were, and situate their weapons/solo status when hitting a kill tile in practice mode. The game will now rescue the tee when it hits a kill tile instead (as if using /rescue). As to not interfere with dummy maps where the team is unlocked & the dummy is killed in quick succession, the tee will only be rescued if the team is locked as well.

## Checklist

- [x] Tested the change ingame
- [ ] Provided screenshots if it is a visual change
- [x] Tested in combination with possibly related configuration options
- [ ] Written a unit test (especially base/) or added coverage to integration test
- [x] Considered possible null pointers and out of bounds array indexing
- [x] Changed no physics that affect existing maps
- [x] Tested the change with [ASan+UBSan or valgrind's memcheck](https://github.com/ddnet/ddnet/#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)

### CODE REVIEW

**Comment by ewancg** [comment]
currently the server sends a msg to everyone in the team that hits a kill tile, should i keep it the way it is, make it per-user or just remove it entirely?

**Comment by Colocolito** [comment]
This is good idea but not at all, since u cant practice anymore some skips that need to skip a killtile... so this idea is just more annoying for some ppl:) idk if its possible but would be nice if u can add a command for active/deactive killtiles in /practice, that would be a good solution.

**Comment by ewancg** [comment]
What

**Comment by Colocolito** [comment]
There are some skips where u just need skip thru killtiles, so i guess u cant  practice to skip them anymore with this?

**Comment by ewancg** [comment]
I don't know what you mean by skipping kill tiles. If you mean clipping through kill tiles, the collision for the kill tiles has not changed at all & so anywhere you would be able to clip through them before, you still can now (even though I'm pretty sure this is impossible, lol). Still not understanding your issue, but if this is a big concern you can now just use an unlocked team

**Comment by Colocolito** [comment]
Ah then its fine if u still can do it, I worried bc i thought u cant do it anymore (ye there are many ways to do it and i usually do, so thats why)

**Comment by ewancg** [comment]
Oh I see. Yeah, you'd have to be going pretty fast but I guess that explains why there's such a thick layer of them outside the map boundaries. And now that I think about it, I have seen someone do it before 😃

**Comment by murpii** [comment]
Slight annoyance: Teleporting onto a kill tile will softlock you and the only option to get out is to kill. Not even /spec works anymore.

**Comment by ewancg** [comment]
Fixed a bug where the rescue point would be set within a kill tile & next time it checks whether to kill or rescue, it rescues you back to within the kill tile, and so on and so forth. This caused a softlock when edging (or teleporting? can't confirm) into a kill tile

**Comment by murpii** [comment]
Seems fine now (both cases)

File: src/game/server/entities/character.cpp [hunk 1]
```diff
@@ -2066,10 +2071,20 @@ void CCharacter::DDRaceTick()
 		}
 	}
 
+	// Don't save if inside a kill tile
+	bool InKillTile = (Collision()->GetCollisionAt(m_Pos.x + GetProximityRadius() / 3.f, m_Pos.y - GetProximityRadius() / 3.f) == TILE_DEATH ||
```
Comment (edg-l) [inline src/game/server/entities/character.cpp:2075]: should have this check in the case rescue is enabled (practice) so we dont do unnecesary work whenver u are not in practice mode (which is most)

Comment (ewancg) [reply src/game/server/entities/character.cpp]: done

**Comment by ewancg** [comment]
should i remove the server option

**Review by edg-l — APPROVED** [review]

**Review by Robyt3 — CHANGES_REQUESTED** [review]
https://github.com/ddnet/ddnet/issues/2232#issuecomment-635982804 and https://github.com/ddnet/ddnet/issues/3024#issuecomment-704328185 need to be addressed before this can get merged. See also https://github.com/ddnet/ddnet/pull/2221#issuecomment-635090073.

**Comment by ewancg** [comment]
fwiw this pr is not prone to any sort of rescue loops except those caused by arrows/speeders. and technically speaking those should also be factored into the rescue logic because it, just like gravity, has the potential to involuntarily kill the player. if we are only allowing rescues to be set where the player is not falling, it makes sense that we should only rescue if the player is not being pushed by an arrow as well

**Review by heinrich5991 — CHANGES_REQUESTED** [review]
I think this feature is a bit hard for a first-time pull request.

The implementation of this shouldn't touch physics code. It should simply enable you do go back to your last safe position after you died, without touching all the code where you could die. Also, see @Robyt3's comment on my general opinion on this.

**Comment by ewancg** [comment]
What are the physics changes exactly

**Comment by heinrich5991** [comment]
The implementation shouldn't touch `HandleSkippableTiles`.

**Comment by ewancg** [comment]
How else do you detect the tee's death as it hits the kill tile?

**Comment by ewancg** [comment]
I suppose I should relay what I already said in the dev channel. I don't think your disagreement with this approach should prevent it from being merged as I've seen it implemented by 2 others and asked for by far more. Typically when people are practicing they want the smallest amount of downtime possible, and adding a tp or kill waiting period would only lengthen it. I am seriously having a hard time imagining a scenario where somebody would want to manually choose to kill *instead* of being rescued. If they want to kill, they will just hit the kill button as quickly as they are rescued.

The whole point is to automate the rescue anyway.

**Comment by heinrich5991** [comment]
> How else do you detect the tee's death as it hits the kill tile?

I have not given it a lot of thought. Generally, a tee is dead after it hit a death tile, so that's a pretty telling sign that it hit one.

> I suppose I should relay what I already said in the dev channel. I don't think your disagreement with this approach should prevent it from being merged as I've seen it implemented by 2 others and asked for by far more.

I think a bad implementation that is likely to stay is worse than no implementation. No implementation means that there's more pressure to create a good implementation.

**Comment by ewancg** [comment]
> I have not given it a lot of thought. Generally, a tee is dead after it hit a death tile, so that's a pretty telling sign that it hit one.
Rescuing after a kill would likely mean a complete overhaul/redefinition of the rescue system. There is no need to over-complicate something like this

> I think a bad implementation that is likely to stay is worse than no implementation. No implementation means that there's more pressure to create a good implementation.

It has been 3 years since the original propositions of this feature. I still don't really understand why you think this is fundamentally flawed anyway. Why is this a bad approach?

**Comment by heinrich5991** [comment]
> > I have not given it a lot of thought. Generally, a tee is dead after it hit a death tile, so that's a pretty telling sign that it hit one.
> 
> Rescuing after a kill would likely mean a complete overhaul/redefinition of the rescue system. There is no need to over-complicate something like this

I think putting rescue code at every possible location where a tee can be killed is not a good way to organize the code. Sometimes, code should be reorganized instead of it being made worse.

**Comment by ewancg** [comment]
I think you misunderstand the purpose of this PR. It's only intended to 'intercept' deaths via. kill tile, an obstacle which can very reasonably get in the way while playing idly. As such, people would want to practice near them. All console & chat kill commands still work in an unadulterated manner. I fail to see how this code is even disorganized anyway; it's simply changing the behavior of a "kill" tile to more accurately represent what a player might want while playing in practice mode. A player can be "killed" in a bunch of other ways, which the rescue command is designed to help with. Don't see how this is different

**Comment by ewancg** [comment]
Bumping this once again. I don't think it makes a lot of sense that after 4 years, 5500 issues/PRs, and multiple iterations of the same idea from different people, heinrich's preference is the only thing keeping this feature from its fruition.

If you asked the people who actually played this game, you'd be hard-pressed to find somebody who doesn't think it's a good idea (or thinks it's better to require manual intervention from the player when the whole point is to make practicing as smooth & quick to recover from as possible). It just seems very weird that /rescue is only intended to recover from certain situations. I think that any interaction with a map hazard should warrant a rescue.

Here we see an issue from deen in support of this feature (#3024).
Here we see a comment from deen which is also in support of this feature ([comment from #2221](https://github.com/ddnet/ddnet/issues/2232#issuecomment-635979607)).
We have an approved review from Ryozuki.
We have a +1 from kristidonk, who (as far as I know) made a GitHub account just to promote this feature.
I have seen many more +1s from people in and outside of GitHub. When this feature is talked about idly I have never heard anyone but heinrich in opposition.

Out of all these positive reviews, why is his able to send the whole thing into limbo? I think this (or #2221) should be merged regardless of heinrich's opinion, especially considering he has not even provided an actual reason for not wanting the game to behave this way, other than [a more technical issue](https://github.com/ddnet/ddnet/pull/2221#issuecomment-636618226) which can be addressed simply & unobtrusively.

**Comment by heinrich5991** [comment]
I have explained how I think this feature should look like. The PR in its current form does not implement it. AFAIK no one has claimed that how I think the feature should look like would make it less useful for players (other than the code not existing yet, obviously).

I think the current form is a hack and should not be merged.

**Pull request closed without merging**